artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RefsOperation.java

```diff
@@ -152,14 +153,14 @@ protected void rollbackRedelivery(Transaction tx, MessageReference ref, long tim
    @Override
    public void afterCommit(final Transaction tx) {
       for (MessageReference ref : refsToAck) {
-         synchronized (ref.getQueue()) {
```

Comment:
if this is removed should the same syn block on the queue in rollback be removed, im not 100% if this is safe though or not, is there a concurrent test case at all around this if not, could one be easily added?

Reply:
I do this because in QueueImpl::postAcknowledge, the only thing that changed queue's state is decrement queue's delivering count, which i believe is thread safe.

While QueueImpl:postRollback is different, it will not return messages if purge on no consumers and there is no consumer at that time, since add/remove consumer will lock the queue. So we must lock the queue when do post rollback.

If no body is 100% sure about this, i will revert this change, since other change has improved a lot.

Reply:
Makes sense. I think best way of getting confidence is a really good test case to prove it

Reply:
hi @michaelandrepearce I get the sync back, i think we can discuss this later, may be another PR.